### DISCUSSION TITLE
Using S3 as Bucket Source and pulling yamls

### DISCUSSION THREAD

**Discussion by DevopsCareer** [fluxcd/flux2#3160] [Q&A] [answered]
I am using S3 as Bucket source, the S3 bucket contains all my yaml specs which I want to deploy on my k8s cluster.

Question: Now I am trying to create a watcher to pull yaml specs from S3 to my cluster which is also running flux-cd , does my source watcher has to use a Kind of type` Kustomization` to use SourceRef as Bucket? Is `Kustomize` is only option to be used as source watchers while pulling yamls from S3 source Bucket?

` sourceRef:`
    `kind: Bucket`

Is there another KindObject which I can use other than Kustomization as source watcher or is there a way I can directly pull from yaml specs from Bucket resource ? I do not want to use Kustomize .

**Comment by stefanprodan** [comment, marked as answer]
You don't have to use Kustomize, see https://fluxcd.io/flux/components/kustomize/kustomization/#generate-kustomizationyaml

**Reply by DevopsCareer** [reply]
Sorry can you point to an example, are you reffering to generating Kustomize automatically ?

**Reply by stefanprodan** [reply]
Flux does it for you, so you don't need to have a `kustomization.yaml` in the Bucket. On the cluster create a Bucket and a Flux Kustomization and that's it.

**Reply by DevopsCareer** [reply]
Thanks I have a bucket already created , so I need another resource just Flux Kustomization like this on cluster?
```
# create kustomization
kustomize create --autodetect --recursive
```

**Reply by kingdonb** [reply]
See `flux create kustomization` which is a totally different Kustomization than the `kustomize` Kustomization

https://fluxcd.io/flux/faq/#are-there-two-kustomization-types

You want `flux create kustomization` to get the other kind 👍

**Reply by DevopsCareer** [reply]
Thanks, that make sense

**Reply by DevopsCareer** [reply]
one subsequent question in line with this, does kustomize.toolkit.fluxcd.io/v1beta2 will be the one I will be using. It seems it does not have `spec.ignore` ? Am I right here that if i have to ignore picking certain file extensions from S3 bucket I have to add the `spec.ignore` only in Source Bucket custom object. If yes , how does relative path works - i tried to give relative path with bucket name as relative path
 eg:
 if bucket name is `"abc"` and I have to ignore path `abc/temp/dir.`
 I just added 
```
 .spec
    .ignore : /temp/dir/*
```
   It did not worked , where does Bucket .ignore starts its path?

**Reply by kingdonb** [reply]
In Flux, ignore is done in the source API, either with a `.sourceignore` file or in `spec.ignore` in the Source. You can read about how it works on the Bucket docs:

https://fluxcd.io/flux/components/source/buckets/

> Storage objects which keys match the defined rules ...

There are a number of methods to add ignores, and there are some defaults to be aware of. The examples here all start with a root slash like yours, and the path which begins after the bucket name. I'm not sure why your wildcard didn't work there, from the docs it looks like it should work. The docs say that ignoring a file should exclude it from fetching as well.

Since the source does handle all the fetching, it would need to be ignored there, rather than in the Kustomization downstream, unless it was going to be needlessly downloaded every time that Flux reconciles the Bucket.

**Reply by DevopsCareer** [reply]
Great thanks, I will review what is wrong with my implementation. Thanks for clarifying

**Comment by kingdonb** [comment]
Source and Workload are two opposite sides of the Flux "GitOps Toolkit" APIs, they don't have to interact.

When you create a `Bucket` source, the Flux source controller is pulling all the content of that bucket into a .tgz artifact or file which gets stored temporarily within the cluster, as a target for pulling from over HTTP.

What you do with that source content is up to you! If you want to apply the manifests to the cluster, you can use a Flux Kustomization to do it, without learning any Kustomize.

If you want to do some other manifest generation (other than Kustomize or not-Kustomize, which are the choices now) we recommend to do it in CI, or with some job that happens ahead of the Source artifact. So that Flux only has to deal with raw Kubernetes YAMLs.
